# Odds model
For our second model, we want to take advantage of the fact that we have the odds calculated by the odds company as part of our dataset. What happens if we just predict that the team with the lowest odds wins the match?

In [213]:
import numpy as np
from utils import DataAggregator

In [214]:
np.random.seed(42)

In [215]:
data_aggregator = DataAggregator()
data = data_aggregator.get_data(["E0"])
data.head()

Div      Date     HomeTeam    AwayTeam  FTHG  FTAG FTR  HTHG  HTAG HTR  ...  \
0  E0  15/08/09  Aston Villa       Wigan     0     2   A     0     1   A  ...   
1  E0  15/08/09    Blackburn    Man City     0     2   A     0     1   A  ...   
2  E0  15/08/09       Bolton  Sunderland     0     1   A     0     1   A  ...   
3  E0  15/08/09      Chelsea        Hull     2     1   H     1     1   D  ...   
4  E0  15/08/09      Everton     Arsenal     1     6   A     0     3   A  ...   

   AF  HC  AC  HY  AY  HR  AR  B365H  B365D  B365A  
0  14   4   6   2   2   0   0   1.67   3.60   5.50  
1   9   5   4   2   1   0   0   3.60   3.25   2.10  
2  10   4   7   2   1   0   0   2.25   3.25   3.25  
3  15  12   4   1   2   0   0   1.17   6.50  21.00  
4  13   4   9   0   0   0   0   3.20   3.25   2.30  

[5 rows x 26 columns]

Now we can start by implementing the base model. This model will like mentioned be based on the betting company's odds calculations for each match. We will simply choose the outcome with the lowest odds, either a home win, an away win, or a draw.

In [ ]:
def predict_result(row):
    if row["B365H"] <= row["B365D"] and row["B365H"] <= row["B365A"]:
        return "H"
    elif row["B365D"] < row["B365H"] and row["B365D"] < row["B365A"]:
        return "D"
    return "A"

data["prediction"] = data.apply(predict_result, axis=1)


In [217]:
accuracy, won = data_aggregator.calculate_accuracy(data, "FTR", "prediction")

In [218]:
print(f"""The accuracy of the model is {accuracy:.2%}
The accuracy of the model is {accuracy*len(data):.0f} out of {len(data)} games.
""")

The accuracy of the model is 54.54%
The accuracy of the model is 4603 out of 8440 games.



In [219]:
print(f"With this model, the expected return of investment would be {won:.2f}€")

With this model, the expected return of investment would be -2464.29€


As we can se from the result above, this model is actually quite accurate, with an accuracy of 54.54%. However, since it always chooses the lowest odds option, the return of investment is negative. This makes sense because even though the model wins 54.54% of its bets, it will rarely get a large return from a single bet. If this model would have yielded a positive return of investment, the betting companies would have had to make a lot of big mistakes in their odds calculations.

In [220]:
data_aggregator.save_metrics("odds", accuracy, won)